flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -369,6 +373,27 @@ private String operatorName(String suffix) {
         default:
           throw new RuntimeException("Unrecognized write.distribution-mode: " + writeMode);
       }
+
+      if (keySelector != null) {
+        return input.keyBy(keySelector);
+      }
+      return input;
+    }
+
+    private KeySelector<RowData, String> getKeySelector(List<Integer> equalityFieldIds, PartitionSpec partitionSpec,
+        Schema schema, RowType rowType) {
+      boolean hasPrimaryKey = equalityFieldIds != null && !equalityFieldIds.isEmpty();
+      boolean hasPartitionKey = partitionSpec != null && !partitionSpec.isUnpartitioned();
+
+      if (hasPrimaryKey && hasPartitionKey) {
+        return new CombinedKeySelector(partitionSpec, equalityFieldIds, schema, rowType);
```

Comment:
this defeats the purpose of hash distribution, which is make sure one writer task processed data from one partition. This `CombinedKeySelector` will nullify the goal. It seems to me that hash shuffle and equality key shuffle shouldn't be enabled at the same time.

Reply:
As far as I know, between equality key and partition key are two independent fields, there is no restriction between them. So I add `CombinedKeySelector` to maintain the independence between them. But that will defeats the purpose of hash distribution as you say. 

restrict enable hash distribution mode when `equalityFieldIds` is not null is one way to reslove this problem, but I think should we restrict partition key must be a subset of equality key to reslove this? Actualy, I have little confuse between equality key and partition key. 

I think partition key should be a subset of equality key, because when user define the equality key in partitioned table that is mean one row should be unique in each partition. Such as we got a table partitioned by <day, hour>, we should ensure equality key contain partition key like <day, hour, id>, but not like <id>. I think the later is confusing and make no sense.
Just like a MySQL table which primary key is <userId>, and sharding data by <day, hour> to different table, and still require keep <userId> is unique in global.

Once we can ensure partition key is a subset of equality key, we can remove the `CombinedKeySelector` and use equality key sheffle data.

Reply:
regardless the answer to the question on "partition key is a subset of equality key", we probably don't need `CombinedKeySelector `
* yes. then equality key shuffle is equivalent to the combined key
* no. we only need the equality key shuffle to maintain the order. combing equality key and partition key shuffle won't bring any benefit of reducing the number of data files (that hash distribution on partition key brings)